x/wasm/IBC.md

```diff
@@ -0,0 +1,253 @@
+should be able to do so).
+
+An "IBC Enabled" contract must support the following callbacks from the runtime
+(we will likely multiplex many over one wasm export, as we do with handle, but these 
+are the different calls we must support):
+
+Packet Lifecycle:
+
+* `IBCRecvPacket` - when another chain sends a packet destined to
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* `IBCPacketAck` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was successfully
+  processed on the other chain (this implies using an envelope for the 
+  IBC acknowledge message, so the Go code can differentiate between success and error)
+* `IBCPacketFailed` - The original sender of `IBCSendMsg` will 
+  get this callback eventually if the message failed to be
+  processed on the other chain (for timeout, closed channel, or the 
+  other side returning an error message in the IBC acknowledgement)
+
+Channel Lifecycle Hooks:
+
+* `OnChanOpenTry` - this is called when another chain attempts to open
+  a channel with this contract. It provides the protocol versions that
+  the initiate expects, and the contract can reject it or accept it
+  and return the protocol version it will communicate with.
+* `OnChanOpenConfirm` - this is called when the other party has agreed
+  to the channel and all negotiation is over. This will be called exactly
+  once for any established channel, and the contract should record these
+  open channels in the internal state.
+* `OnChanClosed` - this is called when an existing channel is closed
+  for any reason, allowing the client to update the state there.
+  This will (likely? @cwgoes?) be accompanied by `IBCPacketFailed`
+  callbacks for all the in-progress packets that were not processed before
+  it closed, as well as all pending acks.
+
+We may want to expose some basic queries of IBC State to the contract.
+We should check if this is useful to the contract and if it opens up
+any possible DoS:
+
+* `ListMyChannels` - return a list of all channel IDs (along with remote PortID)
+  that are bound to this contract's Port.
+* `ListPendingPackets` - given a channelID owned by the contract, return all packets
+  in that channel, that were sent by this chain, for which no acknowledgement or timeout
+  has been received
+
+For mocks, all the `IBCxxxPacket` commands are routed to some
+Golang stub handler, but containing the contract address, so we
+can perform contract-specific actions for each packet.
+
+## Contract Details
+
+Here we map out the workflow with the exact arguments passed with those calls 
+from the Go side (which we will use with our mock), and then a 
+proposal for multiplexing this over fewer wasm exports (define some rust types)
+
+Messages:
+
+```go
+type IBCSendMsg struct {
+    ChannelID string
+    RemotePort string // do we need both? isn't channel enough once it is established???
```

Comment:
>... a single instance of an IBC module may communicate on the same port with any number of other modules and and IBC will correctly route all packets to the relevant module using the (channelID, portID tuple)

https://docs.cosmos.network/master/ibc/overview.html#capabilities

Reply:
@cwgoes can you explain this to me?

I understand **connection** as IP-level connection (the linked docs refers to **channel** as like IP connection)

A channel is like TCP/UDP connection which is established with source and remote ports via handshake. Once the channel is established, it has a well-defined port on both sides and simply the channelID should identify which client to use and the source/dest port, right?

Reply:
Not quite; the terminology is a little different than TCP/IP.

A *connection* exists between two chains and has a unique identifier on each chain. Many channels may use one connection.

A *channel* has a unique identifier per-port, so the combination of port identifier & channel identifier is required to uniquely identify a channel on a single chain. A channel references a single connection, which it will utilise to verify state (that is, it will utilise the client associated with the connection).

Of course, if a contract is always bound to a fixed port, the contract can just identify channels by the channel identifier, and the wasmd runtime can look up the port, presumably.